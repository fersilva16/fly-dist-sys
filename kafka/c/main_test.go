package main

import (
	"gossip-gloomers/testutils"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	maelstrom "github.com/jepsen-io/maelstrom/demo/go"
	"github.com/stretchr/testify/require"
)

func Test(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)
	node1 := testutils.NewClient("n1", link)

	go main()

	err := client.InitNode("n0", []string{"n0", "n1"})

	require.NoError(err)

	sendOutput, err := client.RPC(SendRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "send",
			MsgID: 2,
		},

		Key: "0",
		Msg: 83,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, sendOutput)

	newMessageOutput, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, newMessageOutput)

	pollOutput, err := client.RPC(PollRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "poll",
			MsgID: 3,
		},

		Offsets: map[string]int{
			"0": 0,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, pollOutput)

	sendOutput2, err := client.RPC(SendRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "send",
			MsgID: 4,
		},

		Key: "0",
		Msg: 84,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, sendOutput2)

	newMessageOutput2, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, newMessageOutput2)

	pollOutput2, err := client.RPC(PollRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "poll",
			MsgID: 5,
		},

		Offsets: map[string]int{
			"0": 1,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, pollOutput2)

	commitOffsetsOutput, err := client.RPC(CommitOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "commit_offsets",
			MsgID: 6,
		},

		Offsets: map[string]int{
			"0": 1,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, commitOffsetsOutput)

	commitOffsetsOutput1, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, commitOffsetsOutput1)

	listCommittedOffsetOutput, err := client.RPC(ListCommittedOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "list_committed_offsets",
			MsgID: 7,
		},

		Keys: []string{
			"0",
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, listCommittedOffsetOutput)
}

func TestFollower(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)
	node0 := testutils.NewClient("n0", link)

	go main()

	err := client.InitNode("n1", []string{"n0", "n1"})

	require.NoError(err)

	err = client.Write(SendRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "send",
			MsgID: 2,
		},

		Key: "0",
		Msg: 83,
	})

	require.NoError(err)

	nextOffsetOutput, err := node0.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, nextOffsetOutput)

	err = node0.Write(NextOffsetResponse{
		MessageBody: maelstrom.MessageBody{
			Type:      "next_offset_ok",
			InReplyTo: 1,
		},

		Offset: 0,
	})

	require.NoError(err)

	sendOutput, err := client.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, sendOutput)

	err = node0.Write(NewMessageRequest{
		MessageBody: maelstrom.MessageBody{
			Type: "new_message",
		},

		Key:    "0",
		Offset: 1,
		Msg:    84,
	})

	require.NoError(err)

	pollOutput, err := client.RPC(PollRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "poll",
			MsgID: 4,
		},

		Offsets: map[string]int{
			"0": 0,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, pollOutput)

	commitOffsetsOutput, err := client.RPC(CommitOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "commit_offsets",
			MsgID: 6,
		},

		Offsets: map[string]int{
			"0": 1,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, commitOffsetsOutput)

	commitOffsetsOutput1, err := node0.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, commitOffsetsOutput1)

	listCommittedOffsetOutput, err := client.RPC(ListCommittedOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "list_committed_offsets",
			MsgID: 7,
		},

		Keys: []string{
			"0",
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, listCommittedOffsetOutput)
}
