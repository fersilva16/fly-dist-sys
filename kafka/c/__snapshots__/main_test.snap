
[Test - 1]
{"src":"n0","dest":"c0","body":{"in_reply_to":2,"offset":0,"type":"send_ok"}}
---

[Test - 2]
{"src":"n0","dest":"n1","body":{"type":"new_message","key":"0","offset":0,"msg":83}}
---

[Test - 3]
{"src":"n0","dest":"c0","body":{"in_reply_to":3,"msgs":{"0":[[0,83]]},"type":"poll_ok"}}
---

[Test - 4]
{"src":"n0","dest":"c0","body":{"in_reply_to":4,"offset":1,"type":"send_ok"}}
---

[Test - 5]
{"src":"n0","dest":"n1","body":{"type":"new_message","key":"0","offset":1,"msg":84}}
---

[Test - 6]
{"src":"n0","dest":"c0","body":{"in_reply_to":5,"msgs":{"0":[[1,84]]},"type":"poll_ok"}}
---

[Test - 7]
{"src":"n0","dest":"c0","body":{"in_reply_to":6,"type":"commit_offsets_ok"}}
---

[Test - 8]
{"src":"n0","dest":"n1","body":{"type":"commit_offsets","offsets":{"0":1},"gossip":true}}
---

[Test - 9]
{"src":"n0","dest":"c0","body":{"in_reply_to":7,"offsets":{"0":1},"type":"list_committed_offsets_ok"}}
---

[TestFollower - 1]
{"src":"n1","dest":"n0","body":{"key":"0","msg":83,"msg_id":1,"type":"next_offset"}}
---

[TestFollower - 2]
{"src":"n1","dest":"c0","body":{"in_reply_to":2,"offset":0,"type":"send_ok"}}
---

[TestFollower - 3]
{"src":"n1","dest":"c0","body":{"in_reply_to":4,"msgs":{"0":[[0,83],[1,84]]},"type":"poll_ok"}}
---

[TestFollower - 4]
{"src":"n1","dest":"c0","body":{"in_reply_to":6,"type":"commit_offsets_ok"}}
---

[TestFollower - 5]
{"src":"n1","dest":"n0","body":{"type":"commit_offsets","offsets":{"0":1},"gossip":true}}
---

[TestFollower - 6]
{"src":"n1","dest":"c0","body":{"in_reply_to":7,"offsets":{"0":1},"type":"list_committed_offsets_ok"}}
---
